### PR TITLE
A cleaner and better XYZ (and general I/O) implementation

### PR DESCRIPTION
**Pull request by ceriottm**
Will create a more thorough roadmap in a later comment, but basically
(1) Better handling of units
(2) As efficient as it can get without sacrificing ease of use

### CODE REVIEW

**Comment by ceriottm** [comment]
Changed target for merging to master

**Comment by OndrejMarsalek** [comment]
Note that there are some extra files in this branch that should not be there. I am not able to review the code itself right now, but hopefully soon.

File: ipi/utils/io/backends/io_units.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+traj_dict = Traj().traj_dict                            # trajectory dictionary
+traj_re = [re.compile('%s%s' % (key, '\{[a-z]*\}')) for key in traj_dict.keys()]  # trajectory patterns
+
+def process_units(comment, cell, qatoms, names, masses, output='objects'):
+
+  # Extracting trajectory units
+  family, unit = 'undefined', ''
+  if comment != '':
+    traj = filter(None, [key.search(comment) for key in traj_re])[0].group()[:-1].split('{')
+    family, unit = traj_dict[traj[0]]['dimension'], traj[1]
```
Comment (grhawk) [inline ipi/utils/io/backends/io_units.py]: **WARNING:**

If comment do not contains any matchable string, the program will enter the `if` but `traj` will be an empty list  raising an error in the line below.

Replacing lines 24-25 with the following should fix it.

```python
  is_comment_useful = filter(None, [key.search(comment.strip()) for key in traj_re])
  if len(is_comment_useful) > 0:
    traj = is_comment_useful[0].group()[:-1].split('{')
```

File: ipi/utils/io/backends/io_xyz.py [hunk 2]
```diff
@@ -69,65 +68,70 @@ def print_xyz(atoms, cell, filedesc=sys.stdout, title=""):
         filedesc.write("%8s %12.5e %12.5e %12.5e\n" % (lab[i], qs[3*i], qs[3*i+1], qs[3*i+2]))
 
 
+# Cell type patterns
+cell_re = [re.compile('# CELL.abcABC.: ([-0-9.Ee ]*) '), re.compile('# CELL.GENH.: ([-0-9.Ee ]*)'),
```
Comment (ceriottm) [inline ipi/utils/io/backends/io_xyz.py]: here I am guilty. partially too lazy to type \(, partially wanted to leave some leeway if someone writes CELL[abcABC]. harmless IMO

File: ipi/utils/io/backends/io_xyz.py [hunk 3]
```diff
@@ -69,65 +68,70 @@ def print_xyz(atoms, cell, filedesc=sys.stdout, title=""):
...
-    if reabc is not None:
-        a, b, c, alpha, beta, gamma = reabc.group(1).split()
+    if cell[0] is not None:    # abcABC
+        a, b, c, alpha, beta, gamma = cell[0].group(1).split()
         a = float(a)
         b = float(b)
         c = float(c)
         alpha = float(alpha) * np.pi/180
         beta = float(beta) * np.pi/180
         gamma = float(gamma) * np.pi/180
```
Comment (ceriottm) [inline ipi/utils/io/backends/io_xyz.py]: ok to do np.pi/180.0, actually one could precompute the constant.

File: ipi/utils/io/backends/io_xyz.py [hunk 4]
```diff
@@ -69,65 +68,70 @@ def print_xyz(atoms, cell, filedesc=sys.stdout, title=""):
...
-        y = float(body[2])
-        z = float(body[3])
+    cell = h
+
+    qatoms = np.zeros(3*natoms)
+    names = np.zeros(natoms,dtype='|S4')
+    masses = np.zeros(natoms)
+
+    # Extracting a time-frame information
+    data =list(islice(filedesc,natoms))
```
Comment (ceriottm) [inline ipi/utils/io/backends/io_xyz.py]: Was also thinking of using np.genfromtxt. Unfortunately until python 10 it does not allow to specify how many lines one should read. let me know if you find a good (fast) solution.

File: ipi/engine/atoms.py [hunk 5]
```diff
@@ -259,11 +252,17 @@ def get_kstress(self):
...
 
-      ks = np.zeros((3,3),float)
-      ks[0,0] = np.dot(self.px,self.px/self.m)
-      ks[1,1] = np.dot(self.py,self.py/self.m)
-      ks[2,2] = np.dot(self.pz,self.pz/self.m)
-      ks[0,1] = np.dot(self.px,self.py/self.m)
-      ks[0,2] = np.dot(self.px,self.pz/self.m)
-      ks[1,2] = np.dot(self.py,self.pz/self.m)
+      p = depstrip(self.p)
+      m = depstrip(self.m)
```
Comment (ceriottm) [inline ipi/engine/atoms.py]: No, when you call depstrip it checks for the need of updating. by depstripping you just assume that while you use (without changing!) p and m, they won't be modified elsewhere.

File: ipi/utils/io/backends/io_units.py (hunk 1 above)
Comment (ceriottm) [inline ipi/utils/io/backends/io_units.py]: Seems reasonable, should be tested thorowughlt

Comment (grhawk) [inline ipi/utils/io/backends/io_units.py]: Already done...

File: ipi/engine/atoms.py (hunk 5 above)
Comment (grhawk) [inline ipi/engine/atoms.py]: ok...

File: ipi/utils/io/backends/io_xyz.py (hunk 4 above)
Comment (grhawk) [inline ipi/utils/io/backends/io_xyz.py]: Hello, np.genfromtxt actually has a max_rows option which specify the number of lines that must be read. Unfortunately, stuff containing string must be post-processed anyway...

Comment (ceriottm) [inline ipi/utils/io/backends/io_xyz.py]: Yeah, was referring to the fact that max_rows was introduced in numpy 1.10, and it's not widely available

**Comment by grhawk** [comment]
Sorry, I did not understand.. We could anyway try something like np.fromregex

**Comment by ipoltavskyi** [comment]
Hello everyone,

First of all let me explain why I suggest to use islice method. When just a single line should be read per cycle readline method is obviously faster (approx 2 times).

But with the increase of the number of lines which can be read together slice beats readline and even readlines method. Here are the results of a reading a file which contains 10**7 lines using the following three methods: readlines, deadline, islice.

Readlines method requires 0.974220 (0.056241) sec

Readline method requires 4.740955 (0.320001) sec

Islice method with 1 lines per cycle requires 10.674184 (0.555910) sec
Islice method with 2 lines per cycle requires 5.941821 (0.322189) sec
Islice method with 3 lines per cycle requires 4.261356 (0.201430) sec
Islice method with 4 lines per cycle requires 2.980314 (0.257048) sec
Islice method with 5 lines per cycle requires 2.512731 (0.193435) sec
Islice method with 6 lines per cycle requires 2.098377 (0.139167) sec
Islice method with 7 lines per cycle requires 1.748574 (0.067403) sec
Islice method with 8 lines per cycle requires 1.615319 (0.081661) sec
Islice method with 9 lines per cycle requires 1.556315 (0.098524) sec
Islice method with 50 lines per cycle requires 0.953022 (0.072106) sec
Islice method with 100 lines per cycle requires 0.839183 (0.047002) sec
Islice method with 150 lines per cycle requires 0.901401 (0.213457) sec
Islice method with 200 lines per cycle requires 0.757709 (0.039775) sec
Islice method with 250 lines per cycle requires 0.722690 (0.005129) sec
Islice method with 300 lines per cycle requires 0.719254 (0.006856) sec
Islice method with 350 lines per cycle requires 0.707224 (0.012099) sec
Islice method with 400 lines per cycle requires 0.734456 (0.056408) sec
Islice method with 450 lines per cycle requires 0.964773 (0.054379) sec
Islice method with 500 lines per cycle requires 0.752693 (0.050780) sec
Islice method with 550 lines per cycle requires 0.702607 (0.043130) sec
Islice method with 600 lines per cycle requires 0.674653 (0.007516) sec
Islice method with 650 lines per cycle requires 0.666733 (0.006966) sec
Islice method with 700 lines per cycle requires 0.658762 (0.006628) sec
Islice method with 750 lines per cycle requires 0.658083 (0.005194) sec
Islice method with 800 lines per cycle requires 0.658618 (0.022707) sec
Islice method with 850 lines per cycle requires 0.642924 (0.006391) sec
Islice method with 900 lines per cycle requires 0.629494 (0.006771) sec
Islice method with 950 lines per cycle requires 0.639335 (0.004962) sec

As you can see whenever we read at list 3 lines together islice is more efficient than
readline. Starting from approx 50-100 lines it is considerably faster than even reading an entire 
file with readlines. Since we usually have more than 10 atoms in our simulations islice should be more efficient for i-PI.

Corresponding python script to play with is attached:
[reading_test.py.txt](https://github.com/epfl-cosmo/i-pi-dev/files/487302/reading_test.py.txt)

There are several reasons why islice is more efficient for reading files in our case. First of all we do not need to check EOF condition after reading each line. Second, on a cluster often we are not  the only person who uses the hard drive. Thus, the less times we need to read the better. This is not important for small files, but starting from few GB files it is noticeable. 

In general, after all my efforts to improve the reading methods I found that not really much can be done. A real improvement for post processing can be achieved when I start to use f2py and incorporate fortran functions to compute what I need. For that it is much more efficient to return data as arrays rather than classes. Considerably faster. Changing the way of reading files does not really help much. Thus my personal opinion is that i-PI has grown up enough to learn new languages for post processing such as Fortran.

Regarding reading infrastructure the main problem is correct unit transformation and pdb files. In that case due to the format limitations we can have many troubles even with the existing test cases.

Best regards
Igor

**Comment by ceriottm** [comment]
Thanks Igor, I think we found a way to do this that works as islice and is
much faster than readline(s). However, I agree with you that if one wants
ultimate post-processing speed one should
probably write the whole postproc in fortran.
I suggest we finalize this improved xyz class as well as we can, and then
provide a reasonably fast PPI postprocessor.
Then we can add a contrib/ folder for pure Fortran postprocessor.

**Comment by ipoltavskyi** [comment]
One more potential improvement. There are standard output files such as trajectories.
On the other hand there is no predefined naming scheme. As a result it is hard to write a script which can work in general case. Is it a good idea to add some standart extentions to trajectory files depending upon what is actually inside on io level? Something like ".for."  for forces, ".pos." for positions, etc. In this case it would be much easier to search for the files you need for a particular script. Otherwise for each script a particular naming scheme should be specified. This is a bit inconvenient.

**Comment by ceriottm** [comment]
I hear your pain, but that was sort of by design -- you can for instance
output positions every 10ps, and position of just one bead every ps, so it
would be hard to use a standard naming convention. We need to think this
through a bit.

File: ipi/utils/io/backends/io_xyz.py (hunk 4 above)
Comment (grhawk) [inline ipi/utils/io/backends/io_xyz.py]: image

**Comment by grhawk** [comment]
I think we could merge.... 
even if the io.__init__.read_file could/should be refactored in a nicer way... 

Would that be so bad if we move the `io_units` in a function within the `__init__` and filter the output of the `_get_io_function` directly there?

**Comment by ceriottm** [comment]
I am merging this and deleting the branch. Perhaps @ipoltavskyi can now proceed to include/clean/document the Python version of the PPI estimators, and include also an example of how they are to be used.

**Comment by ipoltavskyi** [comment]
Perfect, thank you Michele.

**Pull request merged**